<a href="https://colab.research.google.com/github/JuanOcampo29/Finanzas/blob/main/Taller_2_Lizarazo_y_Ocampo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Taller 2
###Realizado por Juan Alejandro Ocampo y María Paula Lizarazo Espinosa

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from scipy.optimize import linprog
import cvxpy as cp

## Punto 1

La empresa XYZ produce dos productos: A y B. La ganancia por unidad de A es de 3
dólares y de B es de 5 dólares. Para su producción dispone de 500 unidades de materia
prima, 1000 unidades de mano de obra y 2000 dólares de capital, de los cuales se
necesitan para producir una unidad de:

i) A: una unidad de materia prima, dos unidades de mano de obra y una unidad de capital.

ii) B: una unidad de materia prima, cuatro unidades de mano de obra y dos unidades de
capital.
Además, se necesita producir al menos 100 unidades de A y 150 unidades de B para
satisfacer la demanda del mercado.

Encuentre la asignación óptima de A y B que maximizan las ganancias de la empresa.
Para ello, formule un problema de LP utilizando Scipy.

In [ ]:
tabla_punto_1 = pd.DataFrame({' ' : ['Materia prima', 'Mano de obra', 'Capital','Ganancia'],
        'Producto A' : [1, 2, 1, 3],
        'Producto B' : [1, 4, 2, 5]})
tabla_punto_1

In [ ]:
# Función a maximizar
c = np.array([3,5])

# Restricciones de desigualdad
A = np.array([[-1,0],
              [0,-1],
             [1,1],
             [2,4],
             [1, 2]])

b = np.array([-100,-150,500, 1000, 2000])

# Solución del problema
res = linprog(-c, A_ub=A, b_ub=b, method='simplex') 
res

<ipython-input-4-12ebba9f6dee>:14: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res = linprog(-c, A_ub=A, b_ub=b, method='simplex')


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -1350.0
       x: [ 2.000e+02  1.500e+02]
     nit: 4

Para maximizar las ganancias de la empresa, se dene producir 200 unidades del producto A y 150 del producto B, obteniendo una rentabilidad/ganacia máxima de 1350.

##Punto 2

In [ ]:
tabla_punto_2 = pd.DataFrame({'Valor' : ["50%","30%", "25%","60%"],
        'Tamaño' : ["30%","10%", "40%","20%"],
        'Liquidez' : ["20%","60%", "35%","20%"],
        'E(R)' : ["10%","15%", "16%","8%"]})
tabla_punto_2.index=('Fondo 1', 'Fondo 2', 'Fondo 3','Fondo 4')
tabla_punto_2.T

Fondo 1 Fondo 2 Fondo 3 Fondo 4
Valor        50%     30%     25%     60%
Tamaño       30%     10%     40%     20%
Liquidez     20%     60%     35%     20%
E(R)         10%     15%     16%      8%

Consideramos: 

$$0.5 x_1 + 0.3 x_2 + 0.25 x_3 + 0.6 x_4 ≥ 0.35 (valor)$$

$$0.3 x_1 + 0.1 x_2 + 0.4 x_3 + 0.2 x_4 ≥ 0.30 (tamaño)$$

$$0.2 x_1 + 0.6 x_2 + 0.35 x_3 + 0.2 x_4 ≥ 0.15 (liquidez)$$

$$x_1 + x_2 + x_3 + x_4 = $100$$

$$x_1, x_2, x_3, x_4 ≥ 0$$

In [ ]:
#Con librerías de scipy

# Matriz C
c2 = np.array([0.1,0.15,0.16,0.08])

# Restricciones de igualdad
A2 = np.array([[1,1,1,1]])
b2 = np.array([100])

D2 = np.array([[-0.5,-0.3,-0.25,-0.6],
              [-0.3,-0.1,-0.4,-0.2],
              [-0.2,-0.6,-0.35,-0.2]])
              
d2 = np.array([-35,-30,-15])

# Bounds on decision variables
bounds = [(0,None),
          (0,None),
          (0,None),
          (0,None)]

# Solve the problem:
solution = linprog(-c2, A_eq=A2, b_eq=b2,A_ub=D2, b_ub=d2,
                  bounds=bounds, method='simplex')

solution

<ipython-input-5-0ba926f1914a>:23: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  solution = linprog(-c2, A_eq=A2, b_eq=b2,A_ub=D2, b_ub=d2,


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -13.736842105263158
       x: [ 0.000e+00  1.579e+01  5.789e+01  2.632e+01]
     nit: 7

In [ ]:
#Con cvx
#Número de variables
n=4
x = cp.Variable(n)
objective=cp.Minimize(-c2@x)
#Restricciones
constraints = [A2@x==b2,D2@x<=d2,x>=0,x<=100]
#Planteamiento del problema
prob=cp.Problem(objective,constraints)
#Solución
results=prob.solve()
sol=pd.DataFrame(x.value)
print(round(sol,2))

       0
0   0.00
1  15.79
2  57.89
3  26.32


Se deben invertir 15.79 millones en el fondo 2, 57.89 millones en el fondo 2 y 26.32 millones en el fondo 4 para maximizar la ganancia que, en este caso, es de 13.73 millones

##Punto 3



In [ ]:
tabla_punto_3 = pd.DataFrame({'Mes' : ["","Ene","Feb","Mar","Abr","May","Jun"],
        'FC neto' : ["",-150,-100,200,-200,50,300]})
tabla_punto_3.T

0     1     2    3     4    5    6
Mes         Ene   Feb  Mar   Abr  May  Jun
FC neto    -150  -100  200  -200   50  300

Consideramos: 
$$x_1 + y_1 − z_1 = 150$$
$$x_2+ y_2 −1.01x_1 + 1.003z_1 − z_2 = 100$$
$$x_3 + y_3 −1.01 x_2 + 1.003 z_2 − z_3 = −200$$
$$x_4 − 1.02 y_1 −1.01 x_3 + 1.003 z_3 − z_4 = 200$$
$$x_5 − 1.02 y_2 −1.01 x_4 + 1.003 z_4 − z_5 = −50$$
$$−1.02 y_3 −1.01 x_5 + 1.003 z_5 − z_6 = −300$$

$$x_j ≤ 100 \text {  para}\ j = 1,2,3,4,5$$
$$x_j ≥ 0 \text {  para}\ j = 1,2,3,4,5$$
$$y_j ≥ 0 \text {  para}\ j = 1,2,3$$
$$z_j ≥ 0 \text {  para}\ j = 1,2,3,4,5$$


In [ ]:
#Con librerías de scipy

# Matriz C
c2 = np.array([0,0,0,0,0,0,0,0,0,0,0,0,0,1])

# Restricciones de igualdad
A2 = np.array([[1,0,0,0,0,1,0,0,-1,0,0,0,0,0],
              [-1.01,1,0,0,0,0,1,0,1.003,-1,0,0,0,0],
              [0,-1.01,1,0,0,0,0,1,0,1.003,-1,0,0,0],
              [0,0,-1.01,1,0,-1.02,0,0,0,0,1.003,-1,0,0],
              [0,0,0,-1.01,1,0,-1.02,0,0,0,0,1.003,-1,0],
              [0,0,0,0,-1.01,0,0,-1.02,0,0,0,0,1.003,-1]])
b2 = np.array([150,100,-200,200,-50,-300])

D=np.array([[1,1,1,1,1,0,0,0,0,0,0,0,0,0]])
d=np.array([100])


# Bounds on decision variables
bounds = [(0,100),
          (0,100),
          (0,100),
          (0,100),
          (0,100),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None)]

# Solve the problem:
solution = linprog(-c2, A_eq=A2, b_eq=b2,A_ub=D,b_ub=d,bounds=bounds, method='revised simplex')

solution.x

<ipython-input-29-5d7eb9984dbc>:36: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  solution = linprog(-c2, A_eq=A2, b_eq=b2,A_ub=D,b_ub=d,bounds=bounds, method='revised simplex')


array([  0.        ,  50.98039216,   0.        ,   0.        ,
         0.        , 150.        ,  49.01960784, 203.43436358,
         0.        ,   0.        , 351.9441675 ,   0.        ,
         0.        ,  92.49694915])

In [ ]:
#Con cvx
#Número de variables
n=14
x = cp.Variable(n)
objective=cp.Minimize(-c2@x)
#Restricciones
constraints = [A2@x==b2,x>=0,x[0:5]<=100]
#Planteamiento del problema
prob=cp.Problem(objective,constraints)
#Solución, se usará el método "highs-ipm" para poder comparalo con el codigo de scipy
results2=prob.solve(solver=cp.SCIPY,scipy_options={"method": "highs-ipm"})
solution2=pd.DataFrame(x.value)
print(round(solution2,2))


         0
0    -0.00
1    50.98
2    -0.00
3    -0.00
4    -0.00
5   150.00
6    49.02
7   203.43
8    -0.00
9    -0.00
10  351.94
11   -0.00
12   -0.00
13   92.50


Se invertirá el la línea de crédito solo 50.98, de los 100 que se podían, en el segundo mes; emitirá un papel comercial en enero, febrero y marzo, de 150', 49' y 203' respectivamente. Por último, se evidencia que solo tuvo un saldo para reinvertir en marzo, con un valor de 351.94 millones. De esta manera logra maximizar su saldo en el último mes (92.50)

##Punto 4

La empresa XYZ desea encontrar el saldo óptimo que debe mantener en su cuenta
bancaria en un horizonte de un mes buscando maximizar su rentabilidad total. Se sabe
que la empresa espera recibir entradas de efectivo de:

500 en la primera semana, 3300
en la segunda semana, 200 en la tercera semana y 400 en la cuarta semana. Todas las
cifras se encuentran en millones de pesos. Además, la empresa debe realizar los
siguientes pagos durante el mes: 3350 en la primera semana, 400 en la segunda
semana, 250 en la tercera semana y 200 en la cuarta semana y, para ello, debe
mantener un saldo suficiente que respalde estas obligaciones.

Para aprovechar la rentabilidad que ofrecen los saldos en cada momento, la empresa
puede invertir sus recursos en un instrumento a corto plazo en el que recibe un
rendimiento del 0,02% por día, pero al realizar esta operación debe incurrir en un costo de
transacción del 0,01% del monto transferido.

In [ ]:
tabla_punto_4 = pd.DataFrame({' ' : ['Entradas', 'Salidas', 'Saldos'],
        '1' : [500, 350, 150],
        '2' : [300, 400, -100],
        '3' : [200,250,-50],
        '4' : [400,200,200]})
tabla_punto_4

1    2    3    4
0  Entradas  500  300  200  400
1   Salidas  350  400  250  200
2    Saldos  150 -100  -50  200

In [ ]:
rate=(1+0.0001)**7

# Matriz C
c4 = np.array([0,0,0,0,0,0,0,1])

# Restricciones de igualdad
A4 = np.array([[1,0,0,0,-1,0,0,0],
              [0,1,0,0,rate,-1,0,0],
              [0,0,1,0,0,rate,-1,0],
              [0,0,0,1,0,0,rate,-1]])


b4 = np.array([-150,100,50,-200])


# Solve the problem:
solution = linprog(-c4, A_ub=A4, b_ub=b4,method='simplex')

solution.x

<ipython-input-3-3e57f138ddf3>:17: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  solution = linprog(-c4, A_ub=A4, b_ub=b4,method='simplex')


array([0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       1.50000000e+02, 5.01050315e+01, 1.40115551e-01, 2.00140214e+02])

Como se evidencia, en el último slado se obtiene una rentabilidad de 0.14, pues el saldo para esta semana es de 200.14

##Punto 5

Función a maximizar: 

max es mi función a maximizar

$$ 5x_1+8x_2+10x_3=max$$
$$ x_1+x_2+x_3 = 10000 $$
$$ x_1>=$2000$$
$$ y_2>=300h$$
$$ y_3>=250h$$
$$ y_1+y_2+y_3=1000$$






In [ ]:
c5 = np.array([5,8,10,0,0,0])

A5 = np.array([[1,1,1,0,0,0],
               [0,0,0,1,1,1]])
b5 = np.array([10000,1000])

# Bounds on decision variables
bounds = [(2000,None),
          (2000,None),
          (0,None),
          (0,None),
          (300,None),
          (250,None)]

# Solve the problem:
solution = linprog(-c5, A_eq=A5, b_eq=b5,bounds=bounds, method='simplex')

solution

<ipython-input-9-7909b251ea35>:16: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  solution = linprog(-c5, A_eq=A5, b_eq=b5,bounds=bounds, method='simplex')


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -86000.0
       x: [ 2.000e+03  2.000e+03  6.000e+03  4.500e+02  3.000e+02
            2.500e+02]
     nit: 3

In [ ]:
#Con cvx
#Número de variables
n=6
x = cp.Variable(n)
objective=cp.Minimize(-c5@x)
#Restricciones
constraints = [A5@x==b5,x>=0,x[0:2]>=2000,x[1:4]>=0,x[4]>=300,x[5]>=250]
#Planteamiento del problema
prob=cp.Problem(objective,constraints)
#Solución
results5=prob.solve(solver=cp.SCIPY,scipy_options={"method": "highs-ipm"})
solution5=pd.DataFrame(x.value)
print(round(solution5,2))
#Resolver revisando que el optimizador sea igual a Scipy 

        0
0  2000.0
1  2000.0
2  6000.0
3   450.0
4   300.0
5   250.0


Los presupuestos que se deben asignar a cada sector es de 2000 para ventas, 2000 para marketing y finalmente 6000 para producción. Cabe recalcar que la retricción del departamento de marketing viene dada por el supuesto en el que, como se obtiene mayor ganancia por unidad invertida, el presupuesto que se le asigna debería ser mayor o igual al mínimo de ventas.
De igual manera, se debe invertir 450 horas en ventas, 300 horas en marketing y 250 horas en producción.
Mi ganancia máxima es de 86000

##Punto 6


In [ ]:
tabla_punto_6=pd.DataFrame({"Puntos":[1,2,3,4,5],"Demanda":[80,270,250,160,180]})
tabla_punto_6.T

0    1    2    3    4
Puntos    1    2    3    4    5
Demanda  80  270  250  160  180

In [ ]:
tabla_punto_6_2=pd.DataFrame({"1":["$4","$6","$9"],"2":["$5","$4","$7"],"3":["$6","$3","$4"],
                              "4":["$8","$5","$3"],"5":["$10","$8","$4"],"Fj":["$1000","$1000","$1000"],
                              "Mj":["500","500","500"]})
tabla_punto_6_2.index=("1","2","3")
tabla_punto_6_2

1   2   3   4    5     Fj   Mj
1  $4  $5  $6  $8  $10  $1000  500
2  $6  $4  $3  $5   $8  $1000  500
3  $9  $7  $4  $3   $4  $1000  500

Restricciones
$$ 4x_1+5x_2+6x_3+8x_4+10x_5>=1000$$
$$ 6y_1+4y_2+3y_3+5y_4+8y_5>=1000$$
$$ 9z_1+7z_2+4z_3+3z_4+4z_5>=1000$$
$$x_1+x_2+x_3+x_4+x_5<=500$$
$$y_1+y_2+y_3+y_4+y_5<=500$$
$$z_1+z_2+z_3+z_4+z_5<=500$$
Demanda para las plantas 
$$80+270+250+160+180=940$$
Lo anterior significa que solo necesito dos plantas para satisfacer la demanda de los puntos, razón por la cual, la planta que tenga costos más altos permanecerá cerrada. 
Para esto se definirá una función de linprog poniendo los tres posibles casos, donde la planta 1 o la planta 2 o la planta 3 estén cerradas.

$$x_1 +y_1+z_1>=80$$
$$x_2 +y_2+z_2>=270$$
$$x_3 +y_3+z_3>=250$$
$$x_4+y_4+z_4>=160$$
$$x_5 +y_5+z_5>=180$$

In [ ]:
c6=np.array([4,5,6,8,10,6,4,3,5,8,9,7,4,3,4])

A6=np.array([[-4,-5,-6,-8,-10,0,0,0,0,0,0,0,0,0,0],
             [0,0,0,0,0,-6,-4,-3,-5,-8,0,0,0,0,0],
             [0,0,0,0,0,0,0,0,0,0,-9,-7,-4,-3,-4],
             [1,1,1,1,1,0,0,0,0,0,0,0,0,0,0],
             [0,0,0,0,0,1,1,1,1,1,0,0,0,0,0],
             [0,0,0,0,0,0,0,0,0,0,1,1,1,1,1],
             [-1,0,0,0,0,-1,0,0,0,0,-1,0,0,0,0],
             [0,-1,0,0,0,0,-1,0,0,0,0,-1,0,0,0],
             [0,0,-1,0,0,0,0,-1,0,0,0,0,-1,0,0],
             [0,0,0,-1,0,0,0,0,-1,0,0,0,0,-1,0],
             [0,0,0,0,-1,0,0,0,0,-1,0,0,0,0,-1],
             ])
          
bounds=[(0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None),
          (0,None)]

def min(y):
  solution6 =linprog(c6,A_ub=A6,b_ub=y,bounds=bounds, method='highs-ipm')
  m=pd.DataFrame(solution6.x.reshape(3,5))
  return(solution6.fun)

min(np.array([0,-1000,-1000,0,500,500,-80,-270,-250,-160,-180]))


3610.0

In [ ]:
min(np.array([-1000,0,-1000,500,0,500,-80,-270,-250,-160,-180]))

4050.0

In [ ]:
min(np.array([-1000,-1000,0,500,500,0,-80,-270,-250,-160,-180]))

4840.0

La de menores costos es la opción uno, donde se cierra la planta 1. Cabe recalcar que al costo total falta sumarle los costos fijos del funcionamiento (Puesto que solo hay costos variables de distribución para este ejercicio), por lo que el costo total sería igual a 
$$3610 + 1000 + 1000 = 5610$$

In [ ]:
b6=np.array([0,-1000,-1000,0,500,500,-80,-270,-250,-160,-180])
solution =linprog(c6,A_ub=A6,b_ub=b6,bounds=bounds, method='simplex')
m=pd.DataFrame(solution.x.reshape(3,5))
m

<ipython-input-27-1883ae7bce9d>:2: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  solution =linprog(c6,A_ub=A6,b_ub=b6,bounds=bounds, method='simplex')


0      1      2      3      4
0   0.0    0.0    0.0    0.0    0.0
1  80.0  270.0  150.0    0.0    0.0
2   0.0    0.0  100.0  160.0  180.0

In [ ]:
n=15
x=cp.Variable(n)
objective=cp.Minimize(c6@x)
constraints=[A6@x<=b6,x>=0]
prob=cp.Problem(objective,constraints)
results6=prob.solve(solver=cp.SCIPY,scipy_options={"method": "highs-ipm"})
solution6=pd.DataFrame(x.value.reshape(3,5))
print(round(solution6,2))
results6

      0      1      2      3      4
0  -0.0   -0.0   -0.0   -0.0   -0.0
1  80.0  270.0  150.0   -0.0   -0.0
2  -0.0   -0.0  100.0  160.0  180.0


3610.0

La minimización de los costos se da cerrando la planta 1. La planta 2 satisface la demanda de los puntos 1,2 y 60% del 3; y la planta 3 satisface la demanda de 40% del punto 3, el punto 4 y el punto 5.